Simplify and document running `phpcs` locally

nylen opened the pull request:
This PR makes it easier to run `phpcs` locally.

### To test

- Make sure the Travis build, especially the `phpcs` step, still passes.
- Install `phpcs` following the instructions in the comments of `bin/run-phpcs.sh`.  Run the script and verify that it works (no output, exit code `0`).

nylen committed:
Add a script to run `phpcs` locally

westonruter commented on bin/run-phpcs.sh:
```diff
@@ -0,0 +1,34 @@
+	-not \( -path './node_modules' \) \
+	-not \( -path './vendor' \) \
+	-name '*.php' \
+| xargs -d'\n' phpcs --standard=phpcs.ruleset.xml -s
```
FYI: If `phpcs.ruleset.xml` is renamed to `phpcs.xml.dist` or `phpcs.xml` then you don't have to supply `--standard` at all. I suggest that this could eliminate the need for `bin/run-phpcs.sh` entirely. See https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards#using-a-custom-ruleset

Additionally, the ruleset file can be configured to exclude the `node_modules` and `vendor` directories, eliminating the need for using `find` and `xargs`.

westonruter replied:
The ruleset just needs to include these lines:

```xml
<exclude-pattern>*/node_modules/*</exclude-pattern>
<exclude-pattern>*/vendor/*</exclude-pattern>
```

nylen replied:
Nice! I'll try this out, and then move the setup documentation to a better place.

One more question: is it possible to enable the `-s` option via the config as well? I find it much easier to parse the output this way.

nylen replied:
It turns out we need to whitelist directories and files using the `<file>` tag instead.

With the suggested exclude entries in place, running `phpcs -s .` is extremely, extremely slow (2 minutes and counting before I killed it).  `strace` reveals that `phpcs` still recurses into all directories, ignoring the exclude entries.  This is a known issue, fixed in the 3.x releases, which we can't use yet:  https://github.com/squizlabs/PHP_CodeSniffer/issues/1113

> is it possible to enable the `-s` option via the config as well

From looking through the code, the answer to this is no.

westonruter replied:
From looking at https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml I just found that you can indeed do this! See 7f0fff5.

Now all you need to do to run PHPCS is just do `phpcs` without any args at all.

ntwb reviewed (approved):
Looks good to me 👍 

edit: Ha, I should have refreshed the page first, I would have seen the above ^^^ 😉

nylen committed:
Remove the need for a separate script to run phpcs

- Name the phpcs ruleset with a default filename
- Whitelist PHP files to include in the checks

nylen committed:
Add documentation for running phpcs locally

westonruter committed:
Add default command line args to rulset for -s and --extensions=php

westonruter commented on docs/coding-guidelines.md:
```diff
@@ -78,3 +78,45 @@ Example:
+
+https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards#standalone
+
+You should now be able to run `phpcs -s` from the root directory of this
```
We can now remove ` -s` from the instructions since it is defined as a default arg in `phpcs.xml`

westonruter commented on .travis.yml:
```diff
@@ -58,11 +58,7 @@ script:
-        -not \( -path './vendor' \) \
-        -name '*.php' \
-      | xargs -d'\n' phpcs --standard=phpcs.ruleset.xml -s
+      phpcs -s
```
Per below, the `-s` can now be removed since it is the default.

westonruter commented on docs/coding-guidelines.md:
```diff
@@ -78,3 +78,45 @@ Example:
+not yet compatible with the WordPress coding standards.  For more information see
+[this issue](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/718).
+
+The easiest way to get `phpcs` is to download the .phar archive from the latest
```
I wonder if it would be easier to make `phpcs` 2.9.x a Composer dependency instead? If we add a `composer.json` which has `wp-coding-standards/wpcs` among its dev-requirements then all a contributor (and Travis CI alike) would need to do is `composer install` and then do `./vendor/bin/phpcs`.

nylen replied:
I tried this out, and am getting the following error:

```
PHP Fatal error:  Uncaught exception 'PHP_CodeSniffer_Exception' with message 'Referenced sniff "WordPress-Core" does not exist'
```

Is it expected that we need to call `vendor/bin/phpcs --config-set`?

westonruter replied:
@nylen yes, that's right.

jhabdas replied:
Just landed here from a DDG search for `wordpress exit phpcs` as I'm restructuring my WP plugin a little so I may to also run PHPCS locally. FWIW - I made a suggestion to the WP team earlier today via pull against `wp-cli/sample-plugin` was to make `phpcs` easier to run locally and was shot down gallantly. I'm going to parse through some of this code to look for hints as to what to do in my case. But I was happy to land here and see individuals looking to improve the dev process - because there's room for improvement.

nylen committed:
Remove '-s' argument

nylen commented:
I'll merge this and explore using `composer` in a separate PR.
